### PR TITLE
Refactor: remove Kubo-Greenwood and DOS functions from esolver

### LINKED ISSUE #4344: remove esolver_ks_pw_too.cpp, esolver_sdft_pw_too.cpp and create a new class to do DOS and Kudo-Greenwood calculation
### Describe the Code Quality Issue

DOS and Kubo-Greenwood function are defined in esolver and they should be moved to specific classes.
And esolver will use these new classes.

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Identify the specific code file or section with the code quality issue.
- [ ] Investigate the issue and determine the root cause.
- [ ] Research best practices and potential solutions for the identified issue.
- [ ] Refactor the code to improve code quality, following the suggested solution.
- [ ] Ensure the refactored code adheres to the project's coding standards.
- [ ] Test the refactored code to ensure it functions as expected.
- [ ] Update any relevant documentation, if necessary.
- [ ] Submit a pull request with the refactored code and a description of the changes made.

### PR DESCRIPTION
**Pull request by Qianruipku**
### Linked Issue
Fix #4344

### Unit Tests and/or Case Tests for my changes
- A unit test is added for each new feature or bug fix.

### What's changed?
- mv Kubo-Greenwood to EleCond class
- mv stochastic  Kubo-Greenwood to Sto_EleCond class
- mv DOS of sDFT to Sto_DOS class

### CODE REVIEW

File: source/module_hamilt_pw/hamilt_stodft/sto_dos.cpp [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    const int nk = p_kv->get_nks();
+    Stochastic_Iter& stoiter = p_hsol->stoiter;
+    Stochastic_hchi& stohchi = stoiter.stohchi;
+    const int npwx = p_wfcpw->npwk_max;
+
+    double* spolyv = nullptr;
+    std::complex<double>* allorderchi = nullptr;
+    if (stoiter.method == 1)
+    {
+        spolyv = new double[dos_nche];
```
Comment (mohanchen) [inline source/module_hamilt_pw/hamilt_stodft/sto_dos.cpp]: use std::vector<> instead?

File: source/module_hamilt_pw/hamilt_stodft/sto_dos.cpp [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                int LDA = npwx * nchipk_new * 2;
+                int M = npwx * nchipk_new * 2;
+                dgemm_(&trans, &normal, &N, &N, &M, &kweight, vec_all, &LDA, vec_all, &LDA, &one, spolyv, &N);
+            }
+        }
+    }
+
+    if (stoiter.method == 2)
+    {
+        delete[] allorderchi;
```
Comment (mohanchen) [inline source/module_hamilt_pw/hamilt_stodft/sto_dos.cpp]: use vector?

File: source/module_hamilt_pw/hamilt_stodft/sto_dos.cpp [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    std::ofstream ofsdos;
+    int ndos = int((emax - emin) / de) + 1;
+    stoiter.stofunc.sigma = sigmain / ModuleBase::Ry_to_eV;
+    ModuleBase::timer::tick("Sto_DOS", "Tracepoly");
+
+    std::cout << "2. Dos:" << std::endl;
+    ModuleBase::timer::tick("Sto_DOS", "DOS Loop");
+    int n10 = ndos / 10;
+    int percent = 10;
+    double* sto_dos = new double[ndos];
```
Comment (mohanchen) [inline source/module_hamilt_pw/hamilt_stodft/sto_dos.cpp]: use vector?

Comment (Qianruipku) [reply source/module_hamilt_pw/hamilt_stodft/sto_dos.cpp]: OK

**Review by mohanchen — APPROVED** [review]

**Pull request merged**